subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorsSet.java

```diff
@@ -223,21 +226,36 @@ public Object callOriginal() throws Throwable {
             }
         }
 
-        private InstrumentedGroovyCallsHelper.ThrowingCallable<Object> callableForOriginalCall(
-            CallStrategy callStrategy,
-            Object receiver,
-            @Nullable Object[] args
-        ) {
-            InstrumentedGroovyCallsHelper.ThrowingCallable<Object> result =
-                callStrategy == CallStrategy.CALL_CURRENT ? () -> super.callCurrent((GroovyObject) receiver, args) :
-                    callStrategy == CallStrategy.CALL_GROOVY_OBJECT_GET_PROPERTY ? () -> super.callGroovyObjectGetProperty(receiver) :
-                        null;
-
-            if (result == null) {
-                throw new IllegalArgumentException("unexpected original call strategy " + callStrategy);
+        @SuppressWarnings("DanglingJavadoc")
+        private Object invokeDefaultGroovyCallSiteImplementation(Object receiver, @Nullable Object[] args, CallStrategy callStrategy) throws Throwable {
```

Comment:
💅 I think the structure of this method would be clearer with the case-analysis front and center (and done only once), even if that implies some duplication:

```java
        private Object invokeDefaultGroovyCallSiteImplementation(Object receiver, @Nullable Object[] args, CallStrategy callStrategy) throws Throwable {
            assert groovyDefaultCallSite != this;
            switch (callStrategy) {
                case CALL_CURRENT:
                    if (groovyDefaultCallSite != null) {
                        return groovyDefaultCallSite.callCurrent((GroovyObject) receiver, args);
                    }
                    try {
                        return super.callCurrent((GroovyObject) receiver, args);
                    } finally {
                        restoreCallSite();
                    }
                case CALL_GROOVY_OBJECT_GET_PROPERTY:
                    if (groovyDefaultCallSite != null) {
                        return groovyDefaultCallSite.callGroovyObjectGetProperty(receiver);
                    }
                    try {
                        return groovyDefaultCallSite.callGroovyObjectGetProperty(receiver);
                    } finally {
                        restoreCallSite();
                    }
                default:
                    throw new IllegalArgumentException();
            }
        }
```